File: 4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.py
class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        dp = {}
        ans = float("inf")

        for i, num in enumerate(nums):
            if num not in dp:
                dp[num] = []
            dp[num].append(i)

        for idx in dp.values():
            if len(idx) < 3:
                continue

            for i in range(len(idx) - 2):
                distance = 2 * (idx[i + 2] - idx[i])
                ans = min(ans, distance)

        return -1 if ans == float("inf") else ans

File: 4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumDistance = function (nums) {
    const mp = new Map();

    for (let i = 0; i < nums.length; i++) {
        if (!mp.has(nums[i])) {
            mp.set(nums[i], []);
        }
        mp.get(nums[i]).push(i);
    }

    let ans = Infinity;

    for (const idx of mp.values()) {
        if (idx.length < 3) continue;

        for (let i = 0; i + 2 < idx.length; i++) {
            const distance = 2 * (idx[i + 2] - idx[i]);
            ans = Math.min(ans, distance);
        }
    }

    return ans === Infinity ? -1 : ans;
};